### ISSUE TITLE
Create a file with average values ​from other python files

### ISSUE THREAD

**Issue by demetrafabris** [KratosMultiphysics/Kratos#8797]
I am calculating pressure values ​​at various time steps along 18 lines, and for each line I have 3 rows: one with the pressures at the starting point, one with the pressures at the midpoint and one with the pressures at the end point of each line. I would like to create a dataframe in which for each point (18 x 3 = 54 points) I get the average pressure value for that point. My files (.dat) are like:
```
time  PRESSURE
0.01  150.15
0.02  145.17
0.03  153.25
etc.
```
They are called:` line_output_P1_1, line_output_P1_2, line_output_P1_3, line_output_P2_1, line_output_P2_2, etc.
`
I would like to create a vector with these 54 average pressure values ​​and then plot it with progressive distances in which each distance corresponds to one of the 54 average pressure values

**Comment by loumalouomega** [comment]
What is the problem?, this can be done from python

**Comment by mpentek** [comment]
> Yes, I know, but I don’t know how to do it in Python. I'm not very familiar with python 😅
> […](#)
> Il giorno 20 mag 2021, alle ore 19:11, Vicente Mataix Ferrándiz ***@***.***> ha scritto: ﻿ What is the problem?, this can be done from python — You are receiving this because you authored the thread. Reply to this email directly, view it on GitHub, or unsubscribe.

Your problem is out of scope for the KratosMultiphysics repository. As @loumalouomega pointed out, this is a pure Python technicality. You could also use Matlab, Julia, R or anything else. For such generic questions there are already multitudes of questions and answers posted in dedicated forums. Closing this issue.

**Comment by loumalouomega** [comment]
@demetrafabris I prepared an example for you, it writes the results in a CSV file.
[example.zip](https://github.com/KratosMultiphysics/Kratos/files/6526672/example.zip)

The idea is to create a custom process as the following:

~~~python
# Importing the Kratos Library
import KratosMultiphysics
import os
import csv

def Factory(settings, Model):
    if(type(settings) != KratosMultiphysics.Parameters):
        raise Exception("Expected input shall be a Parameters object, encapsulating a csv string")
    return CustomOutputProcess(Model, settings["Parameters"])

def read_external_csv(output_file_name):
    with open(output_file_name, newline='') as csvfile:
        reader = csv.DictReader(csvfile)
        data = {}
        for row in reader:
            data.update(row)
        return data

def write_external_csv(output_file_name, data):
    add_header = False if os.path.isfile(output_file_name) else True
    with open(output_file_name, 'a', newline='') as csvfile:
        fieldnames = data.keys()
        writer = csv.DictWriter(csvfile, fieldnames=fieldnames)

        if add_header:
            writer.writeheader()
        writer.writerow(data)

# All the processes python processes should be derived from "Process"
class CustomOutputProcess(KratosMultiphysics.Process):
    """This class is used in order to create a csv file containing
    the solution a given model part with a certain frequency

    Only the member variables listed below should be accessed directly.

    Public member variables:
    model -- the model used to construct the process.
    settings -- Kratos parameters containing solver settings.
    """

    def __init__(self, model, params):
        """ The default constructor of the class

        Keyword arguments:
        self -- It signifies an instance of a class.
        model -- the model used to construct the process.
        settings -- Kratos parameters containing solver settings.
        """
        KratosMultiphysics.Process.__init__(self)

        ## Settings string in csv format
        default_parameters = KratosMultiphysics.Parameters("""
        {
            "help"                          : "This process generates a csv file containing the solution of a list of variables from a given submodelpart",
            "output_variables"              : [],
            "output_file_name"              : "",
            "model_part_name"               : "",
            "sub_model_part_name"           : "",
            "check_for_flag"                : "",
            "time_frequency"                : 1.00,
            "historical_value"              : true,
            "resultant_solution"            : false,
            "use_node_coordinates"          : false
        }
        """)

        ## Overwrite the default settings with user-provided parameters
        self.params = params
        self.params.ValidateAndAssignDefaults(default_parameters)

        self.model = model
        self.params = params

        self.time_counter = 0.0

    def ExecuteInitialize(self):
        """ This method is executed at the begining to initialize the process

        Keyword arguments:
        self -- It signifies an instance of a class.
        """

        # We get the submodelpart
        model_part_name = self.params["model_part_name"].GetString()
        sub_model_part_name = self.params["sub_model_part_name"].GetString()
        if sub_model_part_name != "":
            self.sub_model_part = self.model.GetModelPart(model_part_name).GetSubModelPart(sub_model_part_name)
        else:
            self.sub_model_part = self.model.GetModelPart(model_part_name)

        if self.sub_model_part.GetCommunicator().TotalProcesses() > 1: # mpi-execution
            raise Exception("This process cannot be used for writing output in MPI!")

        # If we consider any flag
        flag_name = self.params["check_for_flag"].GetString()
        if flag_name != "":
            self.flag = KratosMultiphysics.KratosGlobals.GetFlag(flag_name)
        else:
            self.flag = None

        self.output_file_name = self.params["output_file_name"].GetString()
        self.output_variables = self.__generate_variable_list_from_input(self.params["output_variables"])
        self.frequency = self.params["time_frequency"].GetDouble()
        self.resultant_solution = self.params["resultant_solution"].GetBool()
        self.historical_value = self.params["historical_value"].GetBool()
        self.use_node_coordinates = self.params["use_node_coordinates"].GetBool()

    def ExecuteBeforeSolutionLoop(self):
        """ This method is executed before starting the time loop

        This step generates the structure of the dictionary

        Keyword arguments:
        self -- It signifies an instance of a class.
        """
        pass

    def ExecuteFinalizeSolutionStep(self):
        """ This method is executed in order to finalize the current step

        Here the dictionary containing the solution is filled

        Keyword arguments:
        self -- It signifies an instance of a class.
        """

        time = self.sub_model_part.ProcessInfo.GetValue(KratosMultiphysics.TIME)
        dt = self.sub_model_part.ProcessInfo.GetValue(KratosMultiphysics.DELTA_TIME)
        self.time_counter += dt
        if self.time_counter > self.frequency:
            data = {}
            self.time_counter = 0.0
            data["TIME"] = time
            count = 0

            # Nodal values
            for node in self.sub_model_part.Nodes:
                compute = self.__check_flag(node)

                if compute:
                    node_identifier = "NODE " + self.__get_node_identifier(node)

                    for i in range(self.params["output_variables"].size()):
                        out = self.params["output_variables"][i]
                        variable_name = out.GetString()
                        variable = KratosMultiphysics.KratosGlobals.GetVariable(variable_name)
                        variable_type = KratosMultiphysics.KratosGlobals.GetVariableType(variable_name)
                        if self.historical_value:
                            value = node.GetSolutionStepValue(variable, 0)
                        else:
                            value = node.GetValue(variable)

                        if variable_type == "Double":
                            if not self.resultant_solution:
                                data[node_identifier + " " + variable_name] = value
                            else:
                                if count == 0:
                                    data["RESULTANT" + " " + variable_name] = value
                                else:
                                    data["RESULTANT" + " " + variable_name] += value
                        elif variable_type == "Array":
                            if KratosMultiphysics.KratosGlobals.GetVariableType(variable_name + "_X") == "Double":
                                if not self.resultant_solution:
                                    data[node_identifier + " " + variable_name + "_X"] = value[0]
                                    data[node_identifier + " " + variable_name + "_Y"] = value[1]
                                    data[node_identifier + " " + variable_name + "_Z"] = value[2]
                                else:
                                    if count == 0:
                                        data["RESULTANT" + " " + variable_name + "_X"] = value[0]
                                        data["RESULTANT" + " " + variable_name + "_Y"] = value[1]
                                        data["RESULTANT" + " " + variable_name + "_Z"] = value[2]
                                    else:
                                        data["RESULTANT" + " " + variable_name + "_X"] += value[0]
                                        data["RESULTANT" + " " + variable_name + "_Y"] += value[1]
                                        data["RESULTANT" + " " + variable_name + "_Z"] += value[2]
                            else:
                                if not self.resultant_solution:
                                    list = self.__kratos_vector_to__python_list(value)
                                    data[node_identifier + " " + variable_name] = list
                                else:
                                    aux = 0.0
                                    for index in range(len(value)):
                                        aux += value[index]
                                    if count == 0:
                                        data["RESULTANT" + " " + variable_name] = aux
                                    else:
                                        data["RESULTANT" + " " + variable_name] += aux
                        elif variable_type == "Vector":
                            if not self.resultant_solution:
                                data[node_identifier + " " + variable_name] = value
                            else:
                                if count == 0:
                                    data["RESULTANT" + " " + variable_name] = value
                                else:
                                    data["RESULTANT" + " " + variable_name] += value

                        # TODO: Add pending classes
                    count += 1

            write_external_csv(self.output_file_name, data)

    def __kratos_vector_to__python_list(self, value):
        """ This method is executed in order to finalize the current computation

        Keyword arguments:
        self -- It signifies an instance of a class.
        value -- The Kratos vector to transform
        """

        list = []
        for index in range(len(value)):
            list.append(value[index])
        return list

    def __generate_variable_list_from_input(self, param):
        """ Parse a list of variables from input.

        Keyword arguments:
        self -- It signifies an instance of a class.
        value -- The Kratos vector to transform
        """

        # At least verify that the input is a string
        if not param.IsArray():
            raise Exception("{0} Error: Variable list is unreadable".format(self.__class__.__name__))

        # Retrieve variable name from input (a string) and request the corresponding C++ object to the kernel
        return [KratosMultiphysics.KratosGlobals.GetVariable(param[i].GetString()) for i in range( 0, param.size())]

    def __check_flag(self, component):
        """ Checks the flag over a component

        Keyword arguments:
        self -- It signifies an instance of a class.
        component -- The Kratos node or element to check
        """

        if self.flag != None:
            if component.Is(self.flag) == False:
                return False

        return True

    def __get_node_identifier(self, node):
        """ returns the identifier/key for saving nodal results in the csv
        this can be either the node Id or its coordinates
        The coordinates can be used to check the nodal results in MPI

        Keyword arguments:
        self -- It signifies an instance of a class.
        node -- The Kratos node to get the identifier for
        """
        if self.use_node_coordinates:
            return 'X_{0:.{digits}f}_Y_{1:.{digits}f}_Z_{2:.{digits}f}'.format(node.X0, node.Y0, node.Z0, digits=6)
        else:
            return str(node.Id)
~~~

You can edit it to adjust to your needs.

and add the following to your Parameters.json:

~~~json
"custom_output_process" : [
{
    "python_module"   : "custom_output_process",
    "help"                  : "",
    "process_name"          : "CustomOutputProcess",
    "Parameters"            : {
        "output_variables"     : ["DISPLACEMENT_X","DISPLACEMENT_Y"],
        "output_file_name"     : "results.csv",
        "model_part_name"      : "Structure.Parts_Parts_Auto1",
        "time_frequency"       : 0.1
    }
}]
~~~
